pkg/pillar/cmd/zedrouter/zedrouter.go

```diff
@@ -871,6 +892,63 @@ func handleAppNetworkCreate(ctxArg interface{}, key string, configArg interface{
 	checkAppNetworkErrorAndStartTimer(ctx)
 }
 
+func publishAppInstMetadata(ctx *zedrouterContext,
+	appInstMetadata *types.AppInstMetaData) {
+	if appInstMetadata == nil {
+		log.Errorf("unpublishAppInstMetadata: nil appInst metadata")
+		return
+	}
+	key := appInstMetadata.Key()
+	log.Functionf("publishAppInstMetadata(%s)", key)
+
+	pub := ctx.pubAppInstMetaData
+	pub.Publish(appInstMetadata.Key(), *appInstMetadata)
+}
+
+func unpublishAppInstMetadata(ctx *zedrouterContext,
+	appInstMetadata *types.AppInstMetaData) {
+	if appInstMetadata == nil {
+		log.Errorf("unpublishAppInstMetadata: nil appInst metadata")
+		return
+	}
+	key := appInstMetadata.Key()
+	log.Tracef("unpublishAppInstMetadata(%s)\n", key)
+
+	pub := ctx.pubAppInstMetaData
+	if exists, _ := pub.Get(key); exists == nil {
+		log.Errorf("unpublishAppInstMetadata(%s) not found\n", key)
+		return
+	}
+	pub.Unpublish(key)
+}
+
+func lookupAppInstMetadata(ctx *zedrouterContext, key string) *types.AppInstMetaData {
+
+	pub := ctx.pubAppInstMetaData
+	st, _ := pub.Get(key)
+	if st == nil {
+		log.Tracef("lookupAppInstMetadata(%s) not found\n", key)
+		return nil
+	}
+	appInstMetadata := st.(types.AppInstMetaData)
+	return &appInstMetadata
+}
+
+func handleAppInstConfigDelete(ctxArg interface{}, key string, configArg interface{}) {
```

Comment:
Can't you do this from the existing AppNetworkConfig delete handler?
Or does zedmanager also delete those when modifying things and not only when we delete the app instance?

Reply:
Seems like there is a bug in zedrouter where it has subscribed for AppNetworkConfig changes from zedagent, where it should have been zedmanager. 
In zedmanager, AppNetworkConfig is unpublished even when we call purge and in such cases, we might not want to delete the metadata. For that reason relying on AppInstConfig's unpublish/delete which is called (if I'm right) only when the appInst is deleted.

Reply:
The AppNetworkCOnfig from zedagent is/was due to the LISP management network. That is dead code today AFAIK.
But zedrouter also subscribes to AppNetworkConfig from zedmanager (for the app instances).

But if zedmanager unpublished on purge then you need this check for AppInstanceConfig that you added.